crates/net/common/src/bandwidth_monitor.rs

```diff
@@ -0,0 +1,277 @@
+
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use reth_ecies::stream::HasRemoteAddr;
+use std::{
+    convert::TryFrom as _,
+    io,
+    net::SocketAddr,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    task::{ready, Context, Poll},
+};
+use tokio::{
+    io::{AsyncRead, AsyncWrite, ReadBuf},
+    net::TcpStream,
+};
+
+/// Meters bandwidth usage of streams
+#[derive(Debug)]
+struct BandwidthMeterInner {
+    /// Measures the number of inbound packets
+    inbound: AtomicU64,
+    /// Measures the number of outbound packets
+    outbound: AtomicU64,
+}
+
+/// Public shareable struct used for getting bandwidth metering info
+#[derive(Clone, Debug)]
+pub struct BandwidthMeter {
+    inner: Arc<BandwidthMeterInner>,
+}
+
+impl BandwidthMeter {
+    /// Returns a new [`BandwidthMeter`].
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(BandwidthMeterInner {
+                inbound: AtomicU64::new(0),
+                outbound: AtomicU64::new(0),
+            }),
+        }
+    }
+
+    /// Returns the total number of bytes that have been downloaded on all the streams.
+    ///
+    /// > **Note**: This method is by design subject to race conditions. The returned value should
+    /// > only ever be used for statistics purposes.
+    pub fn total_inbound(&self) -> u64 {
+        self.inner.inbound.load(Ordering::Relaxed)
+    }
+
+    /// Returns the total number of bytes that have been uploaded on all the streams.
+    ///
+    /// > **Note**: This method is by design subject to race conditions. The returned value should
+    /// > only ever be used for statistics purposes.
+    pub fn total_outbound(&self) -> u64 {
+        self.inner.outbound.load(Ordering::Relaxed)
+    }
+}
+
+impl Default for BandwidthMeter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```

Comment:
this makes the new function a bit redundant, and should even be a clippy lint.

Usually, if `new` does not take any args and returns the type directly (no Result<Self>) then you can just move `new` to `fn default` so you only have 1 way of instantiating the type.

otherwise you have `new` + `default` even though they do the same thing

Reply:
ahh gotcha, appreciate the insight!